### PR TITLE
Register Custom Symbolic of torch.einsum for ORTModule

### PR DESCRIPTION
**Pull request by iK1D**
Support torch.einsum in ORTModule by custom symbolic registration. Current implementation only supports cases in conditions of:
- two operands
- no ellipsis in equation
- no repeated label in single operand (extra torch.diagonal needed for this case)

Below are the exported graphs for some test cases (there is one extra Add before each torch.einsum)

s,se->se: exported to a.unsqueeze([1]) \* b
image

se,sc->sec: exported to a.unsqueeze([2]) \* b.unsqueeze([1])
image

se,se->s: exported to bmm(a.unsqueeze([1]), b.unsqueeze([-1])).reshape([0])
image

sec,sm->ecm: exported to bmm(a.reshape([0,e\*s]).permute([1,0]), b).reshape([e,c,-1])
image

sec,ecm->sm: exported to bmm(a.reshape([0,e\*c]), b.reshape([e\*c,-1]))
image

ks,ksm->sm: exported to bmm(a.unsqueeze([-1]).permute([1,2,0]), b.permute([1,0,2])).reshape([0,-1])
image

kes,ems->kem: exprted to bmm(a.permute([1,0,2]), b.permute([0,2,1])).permute([1,0,2])
image

kes,ksm->ms: exported to bmm(a.sum([1].unsqueeze([-1]).permute([1,2,0]), b.permute([1,0,2])).reshape([0,-1]).permute([1,0])
image

For perf comparison with PyTorch (module with Add+EinSum+MSELOss, K = 16, S = 7840, M = 2048, E = 1024, C = int(S/E*2)):
equation, torch_einsum, torch_composed, ort_einsum, ort_composed
s,se->se: 1.09231ms, 1.06694ms, 1.42507ms, 1.41309ms  // CPU overhead
se,sc->sec: 8.31191ms, 8.14445ms, 8.18224ms, 8.17280ms
se,se->s: 1.26129ms, 1.21959ms, 1.70742ms, 1.71371ms  // CPU overhead + MemCpy(DtoD)
sec,sm->ecm: 72.35454ms, 73.50005ms, 73.72613ms, 73.79965ms  // MemCpy(DtoD) 0.32ms
sec,ecm->sm: 75.63382ms, 74.08072ms, 74.25815ms, 74.16919ms
ks,ksm->sm: 4.57470ms, 4.49915ms, 7.07150ms, 7.07826ms  // Transpose 2.5ms + MemCpy(DtoD) 0.12ms

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py [hunk 1]
```diff
@@ -116,3 +119,252 @@ def avg_pool2d(g, self, kernel_size, stride, padding, ceil_mode, count_include_p
...
+@parse_args('s', 'v')
+def einsum(g, equation, tensor_list):
+    tensors = sym_help._unpack_list(tensor_list)
+    num_ops = len(tensors)
+    assert num_ops > 0
+
+    # Doesn't support implicit output is ellipsis or more than 2 oprands for now.
+    # Doesn't support ellipsis ('...') for now as not easy to get sizes of oprands.
+    if num_ops != 2 or equation.find('->') == -1 or '.' in equation:
+        return g.op("Einsum", *tensors, equation_s=equation)
```
Comment (gramalingam) [inline orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py:213]: Just wondering why we don't implement this as an optimization in ORT (as a graph-rewrite for einsum), which would make it more general? Just trying to understand the reasons, not suggesting doing it now.

Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py:298]: We have the EinSum Op and kernel implemented in ORT for inference, but we don't have the gradient of it for training. This rewrite is needed for training only, and using Python code to implement this here is much easier than using C++ in ORT.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 2]
```diff
@@ -1072,6 +1072,55 @@ def run_step(model, input):
         _test_helpers.assert_values_are_close(ort_prediction, pt_prediction)
         _test_helpers.assert_values_are_close(ort_input.grad, pt_input.grad)
 
+@pytest.mark.parametrize("equation", ["s,se->se", "se,sc->sec", "se,se->s", "sec,sm->ecm",
+                                      "sec,ecm->sm", "ks,ksm->sm", "kes,ems->mek", "kes,ksm->ms"])
+def test_gradient_correctness_einsum(equation):
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1077]: candidates = ["a", "b", "c", "ab", "ba", "ac", "ca", "bc", "cb", "abc", "acb", "bac", "bca", "cab", "cba"]
@pytest.mark.parametrize("left", candidates)
@pytest.mark.parametrize("right", candidates)
def test(left, right): 
    output_pattern = random_generate_a_valid_output_pattern(left, right)
    pattern = "{},{}->{}".format(left, right, output_pattern)

Comment (SherlockNoMad) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1077]: I have tried to read the code, but I don't fully understand the fine details, and I feel there are some tricks that I might have missed. 
I feel that the easiest way to gain some confidence on this piece of the code is to have a good coverage on the test. 
I will be comfortably signing off this PR, if we have a test like above.

Comment (xadupre) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1077]: I did that work for tf2onnx, here is the list of unit tests I put in place to check it was working: https://github.com/onnx/tensorflow-onnx/blob/master/tests/test_einsum_helper.py. It tests many permutation for equations with two terms. I added cases from numpy unit tests.

Comment (iK1D) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1077]: Added a new test function to run such combinations (with some sampling), which will take ~1.5mins. It's good that the "ab*ac->cb" case expose one bug from Gemm-Transpose fusion. I also fixed it in the latest commit.

**Review by SherlockNoMad — APPROVED** [review]

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 2 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1077]: Great! Test coverage paying off.
